// osu-sr-calculator/dist/Objects/Stack.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Stack = void 0;
class Stack {
    constructor(collection) {
        this._size = 0;
        this._version = 0;
        this.DefaultCapacity = 4;
        if (!collection)
            this._array = [];
        else
            this._array = [...collection];
    }
    ;
    get Count() { return this._size; }
    ;
    Clear() {
        this._array.length = 0;
        this._size = 0;
        this._version++;
    }
    ;
    Contains(item) {
        return this._size !== 0 && this._array.lastIndexOf(item, this._size - 1) !== -1;
    }
    ;
    Pop() {
        const size = this._size - 1;
        const array = this._array;
        this._version++;
        this._size = size;
        const item = array[size];
        return item;
    }
    ;
    Push(item) {
        const size = this._size;
        const array = this._array;
        array[size] = item;
        this._version++;
        this._size = size + 1;
    }
    ;
}
exports.Stack = Stack;
;
