// osu-sr-calculator/dist/Objects/osu/HitResult.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.HitResult = void 0;
var HitResult;
(function (HitResult) {
    HitResult[HitResult["None"] = 15] = "None";
    HitResult[HitResult["Miss"] = 5] = "Miss";
    HitResult[HitResult["Meh"] = 4] = "Meh";
    HitResult[HitResult["OK"] = 3] = "OK";
    HitResult[HitResult["Good"] = 2] = "Good";
    HitResult[HitResult["Great"] = 1] = "Great";
    HitResult[HitResult["Perfect"] = 0] = "Perfect";
    HitResult[HitResult["SmallTickMiss"] = 12] = "SmallTickMiss";
    HitResult[HitResult["SmallTickHit"] = 7] = "SmallTickHit";
    HitResult[HitResult["LargeTickMiss"] = 11] = "LargeTickMiss";
    HitResult[HitResult["LargeTickHit"] = 6] = "LargeTickHit";
    HitResult[HitResult["SmallBonus"] = 10] = "SmallBonus";
    HitResult[HitResult["LargeBonus"] = 9] = "LargeBonus";
    HitResult[HitResult["IgnoreMiss"] = 14] = "IgnoreMiss";
    HitResult[HitResult["IgnoreHit"] = 13] = "IgnoreHit";
    HitResult[HitResult["ComboBreak"] = 16] = "ComboBreak";
    HitResult[HitResult["SliderTailHit"] = 8] = "SliderTailHit";
    HitResult[HitResult["LegacyComboIncrease"] = 99] = "LegacyComboIncrease";
})(HitResult || (exports.HitResult = HitResult = {}));
;
