// osu-sr-calculator/dist/Objects/osu/HitWindows.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.GetDifficultyRangeSingle = exports.GetDifficultyRange = exports.HitWindows = exports.DifficultyRange = void 0;
const HitResult_1 = require("./HitResult");
class DifficultyRange {
    constructor(result, min, average, max) {
        this.Min = 0;
        this.Average = 0;
        this.Max = 0;
        this.Result = result;
        this.Min = min;
        this.Average = average;
        this.Max = max;
    }
    ;
}
exports.DifficultyRange = DifficultyRange;
;
class HitWindows {
    constructor() {
        this.base_ranges = [
            new DifficultyRange(HitResult_1.HitResult.Perfect, 22.4, 19.4, 13.9),
            new DifficultyRange(HitResult_1.HitResult.Great, 64, 49, 34),
            new DifficultyRange(HitResult_1.HitResult.Good, 97, 82, 67),
            new DifficultyRange(HitResult_1.HitResult.OK, 127, 112, 97),
            new DifficultyRange(HitResult_1.HitResult.Meh, 151, 136, 121),
            new DifficultyRange(HitResult_1.HitResult.Miss, 188, 173, 158)
        ];
        this.perfect = 0;
        this.great = 0;
        this.good = 0;
        this.ok = 0;
        this.meh = 0;
        this.miss = 0;
    }
    LowestSuccesfulHitResult() {
        for (let result = HitResult_1.HitResult.Meh; result <= HitResult_1.HitResult.Perfect; result++) {
            if (this.IsHitResultAllowed(result))
                return result;
        }
        return HitResult_1.HitResult.None;
    }
    ;
    GetAllAvailableWindows() {
        const res = [];
        for (let result = HitResult_1.HitResult.Meh; result <= HitResult_1.HitResult.Perfect; result++) {
            if (this.IsHitResultAllowed(result))
                res.push({ result: result, length: this.WindowFor(result) });
        }
        return res;
    }
    ;
    IsHitResultAllowed(result) {
        return true;
    }
    ;
    setDifficulty(difficulty) {
        for (const range of this.GetRanges()) {
            const value = (0, exports.GetDifficultyRange)(difficulty, range.Min, range.Average, range.Max);
            switch (range.Result) {
                case HitResult_1.HitResult.Miss:
                    this.miss = value;
                    break;
                case HitResult_1.HitResult.Meh:
                    this.meh = value;
                    break;
                case HitResult_1.HitResult.OK:
                    this.ok = value;
                    break;
                case HitResult_1.HitResult.Good:
                    this.good = value;
                    break;
                case HitResult_1.HitResult.Great:
                    this.great = value;
                    break;
                case HitResult_1.HitResult.Perfect:
                    this.perfect = value;
                    break;
            }
            ;
        }
    }
    ;
    ResultFor(timeOffset) {
        timeOffset = Math.abs(timeOffset);
        for (let result = HitResult_1.HitResult.Perfect; result >= HitResult_1.HitResult.Miss; result--) {
            if (this.IsHitResultAllowed(result) && timeOffset <= this.WindowFor(result))
                return result;
        }
        return HitResult_1.HitResult.None;
    }
    ;
    WindowFor(result) {
        switch (result) {
            case HitResult_1.HitResult.Perfect:
                return this.perfect;
            case HitResult_1.HitResult.Great:
                return this.great;
            case HitResult_1.HitResult.Good:
                return this.good;
            case HitResult_1.HitResult.OK:
                return this.ok;
            case HitResult_1.HitResult.Meh:
                return this.meh;
            case HitResult_1.HitResult.Miss:
                return this.miss;
            default:
                throw new Error(`Unknown hit result: ${result}`);
        }
        ;
    }
    ;
    CanBeHit(timeOffset) {
        return timeOffset <= this.WindowFor(this.LowestSuccesfulHitResult());
    }
    ;
    GetRanges() {
        return this.base_ranges;
    }
    ;
}
exports.HitWindows = HitWindows;
;
const GetDifficultyRange = (difficulty, min, mid, max) => {
    if (difficulty > 5)
        return mid + (max - mid) * (0, exports.GetDifficultyRangeSingle)(difficulty);
    if (difficulty < 5)
        return mid + (mid - min) * (0, exports.GetDifficultyRangeSingle)(difficulty);
    return mid;
};
exports.GetDifficultyRange = GetDifficultyRange;
const GetDifficultyRangeSingle = (difficulty) => (difficulty - 5) / 5;
exports.GetDifficultyRangeSingle = GetDifficultyRangeSingle;
