// osu-sr-calculator/dist/Objects/osu/PathType.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PathType = exports.SplineType = void 0;
var SplineType;
(function (SplineType) {
    SplineType[SplineType["Catmull"] = 0] = "Catmull";
    SplineType[SplineType["Bezier"] = 1] = "Bezier";
    SplineType[SplineType["BSpline"] = 2] = "BSpline";
    SplineType[SplineType["Linear"] = 3] = "Linear";
    SplineType[SplineType["PerfectCurve"] = 4] = "PerfectCurve";
})(SplineType || (exports.SplineType = SplineType = {}));
;
class PathType {
    constructor(splineType, degree = null) {
        this.Degree = null;
        this.Type = splineType;
        this.Degree = null;
    }
    ;
    static { this.CATMULL = new PathType(SplineType.Catmull); }
    static { this.BEZIER = new PathType(SplineType.Bezier); }
    static { this.LINEAR = new PathType(SplineType.Linear); }
    static { this.PERFECT_CURVE = new PathType(SplineType.PerfectCurve); }
    static BSpline(degree) {
        return new PathType(SplineType.BSpline, degree);
    }
    ;
}
exports.PathType = PathType;
;
