// osu-sr-calculator/dist/Objects/osu/HitObjects/SliderObjects/SliderTick.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SliderTick = void 0;
const OsuHitObject_1 = require("../OsuHitObject");
const EmptyHitWindows_1 = require("../EmptyHitWindows");
class SliderTick extends OsuHitObject_1.OsuHitObject {
    constructor(sliderTick) {
        super(sliderTick);
        this.SpanIndex = 0;
        this.SpanStartTime = 0;
        this.PathProgress = 0;
        this.SpanIndex = sliderTick.SpanIndex;
        this.SpanStartTime = sliderTick.SpanStartTime;
        this.PathProgress = sliderTick.PathProgress;
    }
    ;
    ApplyDefaultsToSelf(beatmap) {
        super.ApplyDefaultsToSelf(beatmap);
        let offset;
        if (this.SpanIndex > 0)
            offset = 200;
        else
            offset = this.TimePreempt * 0.66;
        this.TimePreempt = (this.StartTime - this.SpanStartTime) / 2 + offset;
    }
    ;
    CreateHitWindows() {
        return new EmptyHitWindows_1.EmptyHitWindows();
    }
    ;
}
exports.SliderTick = SliderTick;
;
