// osu-sr-calculator/dist/helpers.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.GetPrecisionAdjustedBeatLength = exports.BinarySearch = exports.BinarySearchTimingPoint = exports.lerp = exports.clamp = void 0;
const clamp = (value, min, max) => Math.min(Math.max(value, min), max);
exports.clamp = clamp;
const lerp = (x, y, a) => x * (1 - a) + y * a;
exports.lerp = lerp;
var EqualitySelection;
(function (EqualitySelection) {
    EqualitySelection[EqualitySelection["FirstFound"] = 0] = "FirstFound";
    EqualitySelection[EqualitySelection["Leftmost"] = 1] = "Leftmost";
    EqualitySelection[EqualitySelection["Rightmost"] = 2] = "Rightmost";
})(EqualitySelection || (EqualitySelection = {}));
;
const BinarySearchTimingPoint = (list, time) => {
    let index = (0, exports.BinarySearch)(list, time, EqualitySelection.Rightmost);
    if (index < 0)
        index = ~index - 1;
    return index >= 0 ? list[index] : null;
};
exports.BinarySearchTimingPoint = BinarySearchTimingPoint;
const BinarySearch = (list, time, equalitySelection) => {
    const n = list.length;
    if (n === 0)
        return -1;
    if (time < list[0].Time)
        return -1;
    if (time > list[list.length - 1].Time)
        return ~n;
    let l = 0;
    let r = n - 1;
    let equalityFound = false;
    while (l <= r) {
        const pivot = l + ((r - l) >> 1);
        if (list[pivot].Time < time)
            l = pivot + 1;
        else if (list[pivot].Time > time)
            r = pivot - 1;
        else {
            equalityFound = true;
            switch (equalitySelection) {
                case EqualitySelection.FirstFound:
                    r = pivot - 1;
                    break;
                case EqualitySelection.Leftmost:
                    r = pivot - 1;
                    break;
                case EqualitySelection.Rightmost:
                default:
                    l = pivot + 1;
                    break;
            }
            ;
        }
    }
    if (!equalityFound)
        return ~l;
    switch (equalitySelection) {
        case EqualitySelection.Leftmost:
            return l;
        case EqualitySelection.Rightmost:
        default:
            return l - 1;
    }
    ;
};
exports.BinarySearch = BinarySearch;
const GetPrecisionAdjustedBeatLength = (slider, timingPoint, rulesetShortName) => {
    const sliderVelocityAsBeatLength = -100 / slider.SliderVelocityMultiplier;
    let bpmMultiplier;
    switch (rulesetShortName) {
        case "taiko":
        case "mania":
            bpmMultiplier = sliderVelocityAsBeatLength < 0 ? (0, exports.clamp)(-sliderVelocityAsBeatLength, 10, 10000) / 100 : 1;
            break;
        case "osu":
        case "fruits":
            bpmMultiplier = sliderVelocityAsBeatLength < 0 ? (0, exports.clamp)(-sliderVelocityAsBeatLength, 10, 1000) / 100 : 1;
            break;
    }
    ;
    return timingPoint.BeatLength * bpmMultiplier;
};
exports.GetPrecisionAdjustedBeatLength = GetPrecisionAdjustedBeatLength;
