// osu-sr-calculator/dist/PathApproximator.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PathApproximator = void 0;
const CircularArcProperties_1 = require("./CircularArcProperties");
const Stack_1 = require("./Objects/Stack");
const Vector2_1 = require("./Objects/Vector2");
class PathApproximator {
    static { this.BEZIER_TOLERANCE = 0.25; }
    static { this.catmull_detail = 50; }
    static { this.circular_arc_tolerance = 0.1; }
    static BezierToPiecewiseLinear(controlPoints) {
        return this.BSplineToPiecewiseLinear(controlPoints, Math.max(1, controlPoints.length - 1));
    }
    ;
    static BSplineToPiecewiseLinear(controlPoints, degree) {
        if (controlPoints.length < 2)
            return controlPoints.length === 0 ? [] : [controlPoints[0]];
        degree = Math.min(degree, controlPoints.length - 1);
        const output = [];
        const pointCount = controlPoints.length - 1;
        const toFlatten = this.bSplineToBezierInternal(controlPoints, degree);
        const freeBuffers = new Stack_1.Stack();
        const subdivisionBuffer1 = [];
        const subdivisionBuffer2 = [];
        const leftChild = subdivisionBuffer2;
        while (toFlatten.Count > 0) {
            const parent = toFlatten.Pop();
            if (this.bezierIsFlatEnough(parent)) {
                this.bezierApproximate(parent, output, subdivisionBuffer1, subdivisionBuffer2, degree + 1);
                freeBuffers.Push(parent);
                continue;
            }
            const rightChild = freeBuffers.Count > 0 ? freeBuffers.Pop() : [];
            this.bezierSubdivide(parent, leftChild, rightChild, subdivisionBuffer1, degree + 1);
            for (let i = 0; i < degree + 1; i++) {
                parent[i] = leftChild[i];
            }
            toFlatten.Push(rightChild);
            toFlatten.Push(parent);
        }
        output.push(controlPoints[pointCount]);
        return output;
    }
    ;
    static CatmullToPiecewiseLinear(controlPoints) {
        const result = [];
        for (let i = 0; i < controlPoints.length - 1; i++) {
            const v1 = i > 0 ? controlPoints[i - 1] : controlPoints[i];
            const v2 = controlPoints[i];
            const v3 = i < controlPoints.length - 1 ? controlPoints[i + 1] : v2.add(v2).subtract(v1);
            const v4 = i < controlPoints.length - 2 ? controlPoints[i + 2] : v3.add(v3).subtract(v2);
            for (let c = 0; c < this.catmull_detail; c++) {
                result.push(this.catmullFindPoint(v1, v2, v3, v4, c / this.catmull_detail));
                result.push(this.catmullFindPoint(v1, v2, v3, v4, (c + 1) / this.catmull_detail));
            }
        }
        return result;
    }
    ;
    static CircularArcToPiecewiseLinear(controlPoints) {
        const pr = new CircularArcProperties_1.CircularArcProperties(undefined, controlPoints);
        if (!pr.IsValid)
            return this.BezierToPiecewiseLinear(controlPoints);
        const amountPoints = 2 * pr.Radius <= this.circular_arc_tolerance ? 2 : Math.max(2, Math.ceil(pr.ThetaRange / (2 * Math.acos(1 - this.circular_arc_tolerance / pr.Radius))));
        const output = [];
        for (let i = 0; i < amountPoints; i++) {
            const fract = i / (amountPoints - 1);
            const theta = pr.ThetaStart + pr.Direction * fract * pr.ThetaRange;
            const o = new Vector2_1.Vector2(Math.cos(theta), Math.sin(theta)).scale(pr.Radius);
            output.push(pr.Centre.add(o));
        }
        return output;
    }
    ;
    static LinearToPiecewiseLinear(controlPoints) {
        return controlPoints;
    }
    ;
    static bSplineToBezierInternal(controlPoints, degree) {
        const stack = new Stack_1.Stack();
        degree = Math.min(degree, controlPoints.length - 1);
        const num = controlPoints.length - 1;
        const array = [...controlPoints];
        if (degree === num)
            stack.Push(array);
        else {
            for (let i = 0; i < num - degree; i++) {
                const array2 = [];
                array2[0] = array[i];
                for (let j = 0; j < degree - 1; j++) {
                    array2[j + 1] = array[i + 1];
                    for (let k = 1; k < degree - j; k++) {
                        const num2 = Math.min(k, num - degree - i);
                        array[i + k] = (array[i + k].scale(num2).add(array[i + k + 1])).divide(num2 + 1);
                    }
                }
                array2[degree] = array[i + 1];
                stack.Push(array2);
            }
            stack.Push(array.slice(num - degree));
        }
        return stack;
    }
    ;
    static catmullFindPoint(vec1, vec2, vec3, vec4, t) {
        const t2 = t * t;
        const t3 = t * t2;
        return new Vector2_1.Vector2(0.5 * (2 * vec2.x + (-vec1.x + vec3.x) * t + (2 * vec1.x - 5 * vec2.x + 4 * vec3.x - vec4.x) * t2 + (-vec1.x + 3 * vec2.x - 3 * vec3.x + vec4.x) * t3), 0.5 * (2 * vec2.y + (-vec1.y + vec3.y) * t + (2 * vec1.y - 5 * vec2.y + 4 * vec3.y - vec4.y) * t2 + (-vec1.y + 3 * vec2.y - 3 * vec3.y + vec4.y) * t3));
    }
    ;
    static bezierApproximate(controlPoints, output, subdivisionBuffer1, subdivisionBuffer2, count) {
        const l = subdivisionBuffer2;
        const r = subdivisionBuffer1;
        this.bezierSubdivide(controlPoints, l, r, subdivisionBuffer1, count);
        for (let i = 0; i < count - 1; i++) {
            l[count + i] = r[i + 1];
        }
        output.push(controlPoints[0]);
        for (let i = 1; i < count - 1; i++) {
            const index = 2 * i;
            const p = (l[index - 1].add(l[index].scale(2).add(l[index + 1]))).scale(0.25);
            output.push(p);
        }
    }
    ;
    static bezierSubdivide(controlPoints, l, r, subdivisionBuffer, count) {
        const midpoints = subdivisionBuffer;
        for (let i = 0; i < count; i++) {
            midpoints[i] = controlPoints[i];
        }
        for (let i = 0; i < count; i++) {
            l[i] = midpoints[0];
            r[count - i - 1] = midpoints[count - i - 1];
            for (let j = 0; j < count - i - 1; j++) {
                midpoints[j] = (midpoints[j].add((midpoints[j + 1]))).divide(2);
            }
        }
    }
    ;
    static bezierIsFlatEnough(controlPoints) {
        for (let i = 1; i < controlPoints.length - 1; i++) {
            if ((controlPoints[i - 1].subtract(controlPoints[i].scale(2)).add(controlPoints[i + 1])).lengthSquared() > this.BEZIER_TOLERANCE * this.BEZIER_TOLERANCE * 4) {
                return false;
            }
        }
        return true;
    }
    ;
}
exports.PathApproximator = PathApproximator;
;
exports.default = PathApproximator;
