// osu-sr-calculator/dist/SliderEventGenerator.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SliderEventDescriptor = exports.SliderEventType = exports.SliderEventGenerator = void 0;
const helpers_1 = require("./helpers");
class SliderEventGenerator {
    static { this.TAIL_LENIENCY = -36; }
    static Generate(startTime, spanDuration, velocity, tickDistance, totalDistance, spanCount) {
        const max_length = 100000;
        const length = Math.min(max_length, totalDistance);
        tickDistance = (0, helpers_1.clamp)(tickDistance, 0, length);
        const minDistanceFromEnd = velocity * 10;
        const result = [];
        result.push(new SliderEventDescriptor(SliderEventType.Head, startTime, 0, startTime, 0));
        if (tickDistance !== 0) {
            for (let span = 0; span < spanCount; span++) {
                const spanStartTime = startTime + span * spanDuration;
                const reversed = span % 2 === 1;
                const ticks = this.generateTicks(span, spanStartTime, spanDuration, reversed, length, tickDistance, minDistanceFromEnd);
                if (reversed)
                    ticks.reverse();
                result.push(...ticks);
                if (span < spanCount - 1)
                    result.push(new SliderEventDescriptor(SliderEventType.Repeat, spanStartTime + spanDuration, span, startTime + span * spanDuration, (span + 1) % 2));
            }
        }
        const totalDuration = spanCount * spanDuration;
        const finalSpanIndex = spanCount - 1;
        const finalSpanStartTime = startTime + finalSpanIndex * spanDuration;
        const legacyLastTickTime = Math.max(startTime + totalDuration / 2, finalSpanStartTime + spanDuration + this.TAIL_LENIENCY);
        let legacyLastTickProgress = (legacyLastTickTime - finalSpanStartTime) / spanDuration;
        if (spanCount % 2 === 0)
            legacyLastTickProgress = 1 - legacyLastTickProgress;
        result.push(new SliderEventDescriptor(SliderEventType.LegacyLastTick, legacyLastTickTime, finalSpanIndex, finalSpanStartTime, legacyLastTickProgress));
        result.push(new SliderEventDescriptor(SliderEventType.Tail, startTime + totalDuration, finalSpanIndex, startTime + (spanCount - 1) * spanDuration, spanCount % 2));
        return result;
    }
    ;
    static generateTicks(spanIndex, spanStartTime, spanDuration, reversed, length, tickDistance, minDistanceFromEnd) {
        const result = [];
        for (let d = tickDistance; d <= length; d += tickDistance) {
            if (d >= length - minDistanceFromEnd)
                break;
            const pathProgress = d / length;
            const timeProgress = reversed ? 1 - pathProgress : pathProgress;
            result.push(new SliderEventDescriptor(SliderEventType.Tick, spanStartTime + timeProgress * spanDuration, spanIndex, spanStartTime, pathProgress));
        }
        return result;
    }
    ;
}
exports.SliderEventGenerator = SliderEventGenerator;
;
var SliderEventType;
(function (SliderEventType) {
    SliderEventType[SliderEventType["Tick"] = 0] = "Tick";
    SliderEventType[SliderEventType["LegacyLastTick"] = 1] = "LegacyLastTick";
    SliderEventType[SliderEventType["Head"] = 2] = "Head";
    SliderEventType[SliderEventType["Tail"] = 3] = "Tail";
    SliderEventType[SliderEventType["Repeat"] = 4] = "Repeat";
})(SliderEventType || (exports.SliderEventType = SliderEventType = {}));
;
class SliderEventDescriptor {
    constructor(type, time, spanIndex, spanStartTime, pathProgress) {
        this.Time = 0;
        this.SpanIndex = 0;
        this.SpanStartTime = 0;
        this.PathProgress = 0;
        this.Type = type;
        this.Time = time;
        this.SpanIndex = spanIndex;
        this.SpanStartTime = spanStartTime;
        this.PathProgress = pathProgress;
    }
    ;
}
exports.SliderEventDescriptor = SliderEventDescriptor;
;
