// osu-sr-calculator/dist/SliderPath.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.SliderPath = void 0;
const PathType_1 = require("./Objects/osu/PathType");
const Vector2_1 = require("./Objects/Vector2");
const Precision_1 = require("./Precision");
const Cache_1 = require("./Objects/Cache");
const PathApproximator_1 = __importDefault(require("./PathApproximator"));
const CircularArcProperties_1 = require("./CircularArcProperties");
const helpers_1 = require("./helpers");
class SliderPath {
    get ExpectedDistance() { return this.expectedDistance; }
    ;
    set ExpectedDistance(value) {
        if (value !== this.ExpectedDistance) {
            this.expectedDistance = value;
            this.invalidate();
        }
    }
    ;
    get HasValidLength() { return Precision_1.Precision.DefinitelyBigger(this.Distance, 0); }
    ;
    get ControlPoints() { return this.controlPoints; }
    ;
    set ControlPoints(value) {
        if (value.length) {
            this.controlPoints.push(...value);
            this.invalidate();
        }
    }
    ;
    constructor(sliderPath) {
        this.Version = 0;
        this.expectedDistance = 0;
        this.controlPoints = [];
        this.calculatedPath = [];
        this.cumulativeLength = [];
        this.pathCache = new Cache_1.Cache();
        this.optimisedLength = 0;
        this.calculatedLength = 0;
        this.segmentEnds = [];
        this.segmentEndDistances = [];
        this.ExpectedDistance = sliderPath.ExpectedDistance;
        this.ControlPoints = sliderPath.ControlPoints;
    }
    ;
    get Distance() {
        this.ensureValid();
        return this.cumulativeLength.length === 0 ? 0 : this.cumulativeLength[this.cumulativeLength.length - 1];
    }
    ;
    get CalculatedDistance() {
        this.ensureValid();
        return this.calculatedLength;
    }
    ;
    get OptimiseCatmull() { return this.optimiseCatmull; }
    ;
    set OptimiseCatmull(value) {
        this.optimiseCatmull = value;
        this.invalidate();
    }
    ;
    GetPathToProgress(path, p0, p1) {
        this.ensureValid();
        const d0 = this.progressToDistance(p0);
        const d1 = this.progressToDistance(p1);
        path.length = 0;
        let i = 0;
        for (; i < this.calculatedPath.length && this.cumulativeLength[i] < d0; i++) {
            // ???? quality osu code
        }
        path.push(this.interpolateVertices(i, d0));
        for (; i < this.calculatedPath.length && this.cumulativeLength[i] <= d1; i++) {
            path.push(this.calculatedPath[i]);
        }
        path.push(this.interpolateVertices(i, d1));
    }
    ;
    PositionAt(progress) {
        this.ensureValid();
        const d = this.progressToDistance(progress);
        return this.interpolateVertices(this.indexOfDistance(d), d);
    }
    ;
    PointsInSegment(controlPoint) {
        let found = false;
        const pointsInCurrentSegment = [];
        for (const point of this.ControlPoints) {
            if (!point.Type) {
                if (!found)
                    pointsInCurrentSegment.length = 0;
                else {
                    pointsInCurrentSegment.push(point);
                    break;
                }
            }
            pointsInCurrentSegment.push(point);
            if (point.ToString() === controlPoint.ToString())
                found = true;
        }
        return pointsInCurrentSegment;
    }
    ;
    GetSegmentEnds() {
        this.ensureValid();
        return this.segmentEndDistances.map(d => d / this.Distance);
    }
    ;
    invalidate() {
        this.pathCache.Invalidate();
        this.Version++;
    }
    ;
    ensureValid() {
        if (this.pathCache.IsValid)
            return;
        this.calculatePath();
        this.calculateLength();
        this.pathCache.Validate();
    }
    ;
    calculatePath() {
        this.calculatedPath.length = 0;
        this.segmentEnds.length = 0;
        this.optimisedLength = 0;
        if (this.ControlPoints.length === 0)
            return;
        const vertices = this.ControlPoints.map(p => p.Position);
        let start = 0;
        for (let i = 0; i < this.ControlPoints.length; i++) {
            if (!this.ControlPoints[i].Type && i < this.ControlPoints.length - 1)
                continue;
            const segmentVertices = vertices.slice(start, i + 1);
            const segmentType = this.ControlPoints[start].Type ?? PathType_1.PathType.LINEAR;
            if (segmentVertices.length === 1)
                this.calculatedPath.push(segmentVertices[0]);
            else if (segmentVertices.length > 1) {
                const subPath = this.calculateSubPath(segmentVertices, segmentType);
                const skipFirst = this.calculatedPath.length > 0 && subPath.length > 0 && this.calculatedPath[this.calculatedPath.length - 1].equals(subPath[0]);
                for (let j = skipFirst ? 1 : 0; j < subPath.length; j++) {
                    this.calculatedPath.push(subPath[j]);
                }
            }
            if (i > 0)
                this.segmentEnds.push(this.calculatedPath.length - 1);
            start = i;
        }
        ;
    }
    ;
    calculateSubPath(subControlPoints, type) {
        switch (type.Type) {
            case PathType_1.SplineType.Linear:
                return PathApproximator_1.default.LinearToPiecewiseLinear(subControlPoints);
            case PathType_1.SplineType.PerfectCurve: {
                if (subControlPoints.length !== 3)
                    break;
                const circularArcProperties = new CircularArcProperties_1.CircularArcProperties(undefined, subControlPoints);
                if (!circularArcProperties.IsValid)
                    break;
                const subPoints = (2 * circularArcProperties.Radius <= 0.1) ? 2 : Math.max(2, Math.ceil(circularArcProperties.ThetaRange / (2 * Math.acos(1 - (0.1 / circularArcProperties.Radius)))));
                if (subPoints >= 1000)
                    break;
                const subPath = PathApproximator_1.default.CircularArcToPiecewiseLinear(subControlPoints);
                if (subPath.length === 0)
                    break;
                return subPath;
            }
            case PathType_1.SplineType.Catmull: {
                const subPath = PathApproximator_1.default.CatmullToPiecewiseLinear(subControlPoints);
                if (!this.OptimiseCatmull)
                    return subPath;
                const optimisedPath = [];
                let lastStart = null;
                let lengthRemovedSinceStart = 0;
                for (let i = 0; i < subPath.length; i++) {
                    if (!lastStart) {
                        optimisedPath.push(subPath[i]);
                        lastStart = subPath[i];
                        continue;
                    }
                    const distFromStart = lastStart.distance(subPath[i]);
                    lengthRemovedSinceStart += subPath[i - 1].distance(subPath[i]);
                    const catmull_detail = 50;
                    const catmull_segment_length = catmull_detail * 2;
                    if (distFromStart > 6 || (i + 1) % catmull_segment_length === 0 || i === subPath.length - 1) {
                        optimisedPath.push(subPath[i]);
                        this.optimisedLength += lengthRemovedSinceStart - distFromStart;
                        lastStart = null;
                        lengthRemovedSinceStart = 0;
                    }
                }
                return optimisedPath;
            }
        }
        return PathApproximator_1.default.BSplineToPiecewiseLinear(subControlPoints, type.Degree ?? subControlPoints.length);
    }
    ;
    calculateLength() {
        this.calculatedLength = this.optimisedLength;
        this.cumulativeLength.length = 0;
        this.cumulativeLength.push(0);
        for (let i = 0; i < this.calculatedPath.length - 1; i++) {
            const diff = this.calculatedPath[i + 1].subtract(this.calculatedPath[i]);
            this.calculatedLength += diff.length();
            this.cumulativeLength.push(this.calculatedLength);
        }
        this.segmentEndDistances = [];
        for (let i = 0; i < this.segmentEnds.length; i++) {
            this.segmentEndDistances[i] = this.cumulativeLength[this.segmentEnds[i]];
        }
        if (this.calculatedLength !== this.ExpectedDistance) {
            if (this.calculatedPath.length >= 2 && this.calculatedPath[this.calculatedPath.length - 1].equals(this.calculatedPath[this.calculatedPath.length - 2]) && this.ExpectedDistance > this.calculatedLength) {
                this.cumulativeLength.push(this.calculatedLength);
                return;
            }
            this.cumulativeLength.splice(this.cumulativeLength.length - 1, 1);
            let pathEndIndex = this.calculatedPath.length - 1;
            if (this.calculatedLength > this.ExpectedDistance) {
                while (this.cumulativeLength.length > 0 && this.cumulativeLength[this.cumulativeLength.length - 1] >= this.ExpectedDistance) {
                    this.cumulativeLength.splice(this.cumulativeLength.length - 1, 1);
                    this.calculatedPath.splice(pathEndIndex--, 1);
                }
            }
            if (pathEndIndex <= 0) {
                this.cumulativeLength.push(0);
                return;
            }
            const dir = (this.calculatedPath[pathEndIndex].subtract(this.calculatedPath[pathEndIndex - 1])).normalize();
            this.calculatedPath[pathEndIndex] = this.calculatedPath[pathEndIndex - 1].add(dir.scale(this.ExpectedDistance - this.cumulativeLength[this.cumulativeLength.length - 1]));
            this.cumulativeLength.push(this.ExpectedDistance);
        }
    }
    ;
    MirrorPath() {
        this.calculatedPath.forEach(p => {
            p.y = -p.y;
        });
    }
    ;
    progressToDistance(progress) {
        return (0, helpers_1.clamp)(progress, 0, 1) * this.Distance;
    }
    ;
    interpolateVertices(i, d) {
        if (this.calculatedPath.length === 0)
            return new Vector2_1.Vector2(0, 0);
        if (i <= 0)
            return this.calculatedPath[0];
        if (i >= this.calculatedPath.length)
            return this.calculatedPath[this.calculatedPath.length - 1];
        const p0 = this.calculatedPath[i - 1];
        const p1 = this.calculatedPath[i];
        const d0 = this.cumulativeLength[i - 1];
        const d1 = this.cumulativeLength[i];
        if (Precision_1.Precision.almostEqualsNumber(d0, d1))
            return p0;
        const w = (d - d0) / (d1 - d0);
        return p0.add(p1.subtract(p0).scale(w));
    }
    ;
    indexOfDistance(d) {
        let i = this.cumulativeLength.indexOf(d);
        if (i < 0)
            i = ~this.cumulativeLength.findIndex(c => c > d);
        if (i < 0)
            i = ~i;
        return i;
    }
    ;
}
exports.SliderPath = SliderPath;
;
