// osu-sr-calculator/dist/Objects/osu/HitObjects/DifficultyHitObject.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DifficultyHitObject = void 0;
/**
 * HitObject with additional difficulty properties required to calculate the star rating
 */
class DifficultyHitObject {
    constructor(hitObject, lastObject, clockRate, objects, index) {
        this.difficultyHitObjects = objects;
        this.Index = index;
        this.BaseObject = hitObject;
        this.LastObject = lastObject;
        this.DeltaTime = (hitObject.StartTime - lastObject.StartTime) / clockRate;
        this.StartTime = hitObject.StartTime / clockRate;
        this.EndTime = hitObject.GetEndTime() / clockRate;
    }
    ;
    Previous(backwardsIndex) {
        const index = this.Index - (backwardsIndex + 1);
        return index >= 0 && index < this.difficultyHitObjects.length ? this.difficultyHitObjects[index] : null;
    }
    ;
    Next(forwardsIndex) {
        const index = this.Index + (forwardsIndex + 1);
        return index >= 0 && index < this.difficultyHitObjects.length ? this.difficultyHitObjects[index] : null;
    }
    ;
}
exports.DifficultyHitObject = DifficultyHitObject;
;
