// osu-sr-calculator/dist/Objects/osu/HitObjects/HitObject.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.HitObject = void 0;
const lodash_1 = require("lodash");
const HitWindows_1 = require("../HitWindows");
/**
 * Base class for all osu! hit objects
 */
class HitObject {
    constructor(hitObject) {
        this.control_point_leniency = 1;
        this.StartTime = 0;
        this.NestedHitObjects = [];
        this.StartTime = hitObject.StartTime;
    }
    ;
    ApplyDefaults(beatmap) {
        this.ApplyDefaultsToSelf(beatmap);
        this.NestedHitObjects.length = 0;
        this.CreateNestedHitObjects();
        this.NestedHitObjects.sort((a, b) => a.StartTime - b.StartTime);
        for (const h of this.NestedHitObjects) {
            h.ApplyDefaults(beatmap);
        }
    }
    ;
    ApplyDefaultsToSelf(beatmap) {
        this.HitWindows = this.CreateHitWindows();
        this.HitWindows?.setDifficulty(beatmap.Difficulty.OverallDifficulty);
    }
    ;
    GetEndTime() {
        return (0, lodash_1.round)(this.EndTime ?? this.StartTime, 0);
    }
    ;
    CreateHitWindows() {
        return new HitWindows_1.HitWindows();
    }
    ;
    CreateNestedHitObjects() {
    }
    ;
    AddNested(hitObject) {
        this.NestedHitObjects.push(hitObject);
    }
    ;
}
exports.HitObject = HitObject;
;
