// osu-sr-calculator/dist/Objects/osu/HitObjects/OsuDifficultyHitObject.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.OsuDifficultyHitObject = exports.MIN_DELTA_TIME = exports.NORMALISED_DIAMETER = exports.NORMALISED_RADIUS = void 0;
const helpers_1 = require("../../../helpers");
const HitResult_1 = require("../HitResult");
const DifficultyHitObject_1 = require("./DifficultyHitObject");
const Slider_1 = require("./Slider");
const SliderRepeat_1 = require("./SliderObjects/SliderRepeat");
const SliderTick_1 = require("./SliderObjects/SliderTick");
const Spinner_1 = require("./Spinner");
exports.NORMALISED_RADIUS = 50;
exports.NORMALISED_DIAMETER = exports.NORMALISED_RADIUS * 2;
exports.MIN_DELTA_TIME = 25;
class OsuDifficultyHitObject extends DifficultyHitObject_1.DifficultyHitObject {
    constructor(hitObject, lastObject, clockRate, objects, index, lastLastObject) {
        super(hitObject, lastObject, clockRate, objects, index);
        this.maximum_slider_radius = exports.NORMALISED_RADIUS * 2.4;
        this.assumed_slider_radius = exports.NORMALISED_RADIUS * 1.8;
        this.LazyJumpDistance = 0;
        this.MinimumJumpDistance = 0;
        this.MinimumJumpTime = 0;
        this.TravelDistance = 0;
        this.TravelTime = 0;
        this.Angle = null;
        this.HitWindowGreat = 0;
        this.lastLastObject = lastLastObject;
        this.lastObject = lastObject;
        this.StrainTime = Math.max(this.DeltaTime, exports.MIN_DELTA_TIME);
        if (this.BaseObject instanceof Slider_1.Slider) {
            const sliderObject = this.BaseObject;
            this.HitWindowGreat = 2 * sliderObject.HeadCircle.HitWindows.WindowFor(HitResult_1.HitResult.Great) / clockRate;
        }
        else
            this.HitWindowGreat = 2 * this.BaseObject.HitWindows.WindowFor(HitResult_1.HitResult.Great) / clockRate;
        this.setDistances(clockRate);
    }
    ;
    setDistances(clockRate) {
        if (this.BaseObject instanceof Slider_1.Slider) {
            const currentSlider = this.BaseObject;
            this.computeSliderCursorPosition(currentSlider);
            this.TravelDistance = currentSlider.LazyTravelDistance * Math.pow(1 + currentSlider.RepeatCount / 2.5, 1 / 2.5);
            this.TravelTime = Math.max(currentSlider.LazyTravelTime / clockRate, exports.MIN_DELTA_TIME);
        }
        if (this.BaseObject instanceof Spinner_1.Spinner || this.lastObject instanceof Spinner_1.Spinner)
            return;
        let scalingFactor = exports.NORMALISED_RADIUS / this.BaseObject.Radius;
        if (this.BaseObject.Radius < 30) {
            const smallCircleBonus = Math.min(30 - this.BaseObject.Radius, 5) / 50;
            scalingFactor *= 1 + smallCircleBonus;
        }
        const lastCursorPosition = this.getEndCursorPosition(this.lastObject);
        this.LazyJumpDistance = (this.BaseObject.StackedPosition.scale(scalingFactor).subtract(lastCursorPosition.scale(scalingFactor))).length();
        this.MinimumJumpTime = this.StrainTime;
        this.MinimumJumpDistance = this.LazyJumpDistance;
        if (this.lastObject instanceof Slider_1.Slider) {
            const lastSlider = this.lastObject;
            const lastTravelTime = Math.max(lastSlider.LazyTravelTime / clockRate, exports.MIN_DELTA_TIME);
            this.MinimumJumpTime = Math.max(this.StrainTime - lastTravelTime, exports.MIN_DELTA_TIME);
            const tailJumpDistance = lastSlider.TailCircle.StackedPosition.subtract(this.BaseObject.StackedPosition).length() * scalingFactor;
            this.MinimumJumpDistance = Math.max(0, Math.min(this.LazyJumpDistance - (this.maximum_slider_radius - this.assumed_slider_radius), tailJumpDistance - this.maximum_slider_radius));
        }
        if (!!this.lastLastObject && !(this.lastLastObject instanceof Spinner_1.Spinner)) {
            const lastLastCursorPosition = this.getEndCursorPosition(this.lastLastObject);
            const v1 = lastLastCursorPosition.subtract(this.lastObject.StackedPosition);
            const v2 = this.BaseObject.StackedPosition.subtract(lastCursorPosition);
            let dot = v1.dot(v2);
            let det = v1.x * v2.y - v1.y * v2.x;
            this.Angle = Math.abs(Math.atan2(det, dot));
        }
    }
    ;
    computeSliderCursorPosition(slider) {
        if (!!slider.LazyEndPosition)
            return;
        let trackingEndTime = Math.max(slider.StartTime + slider.Duration - 36, slider.StartTime + slider.Duration / 2);
        let nestedHitObjects = slider.NestedHitObjects;
        let lastRealTick = null;
        for (const hitObject of slider.NestedHitObjects) {
            if (hitObject instanceof SliderTick_1.SliderTick)
                lastRealTick = hitObject;
        }
        if (lastRealTick?.StartTime > trackingEndTime) {
            trackingEndTime = lastRealTick.StartTime;
            const reordered = nestedHitObjects.filter(o => o.StartTime !== lastRealTick.StartTime);
            reordered.push(lastRealTick);
            nestedHitObjects = reordered;
        }
        slider.LazyTravelTime = trackingEndTime - slider.StartTime;
        let endTimeMin = slider.LazyTravelTime / slider.SpanDuration;
        if (endTimeMin % 2 >= 1)
            endTimeMin = 1 - endTimeMin % 1;
        else
            endTimeMin = endTimeMin % 1;
        slider.LazyEndPosition = slider.StackedPosition.add(slider.Path.PositionAt(endTimeMin));
        let currCursorPosition = slider.StackedPosition;
        const scalingFactor = exports.NORMALISED_RADIUS / slider.Radius;
        for (let i = 1; i < nestedHitObjects.length; i++) {
            const currMovementObj = nestedHitObjects[i];
            let currMovement = currMovementObj.StackedPosition.subtract(currCursorPosition);
            let currMovementLength = scalingFactor * currMovement.length();
            let requiredMovement = this.assumed_slider_radius;
            if (i === nestedHitObjects.length - 1) {
                const lazyMovement = slider.LazyEndPosition.subtract(currCursorPosition);
                if (lazyMovement.length() < currMovement.length())
                    currMovement = lazyMovement;
                currMovementLength = scalingFactor * currMovement.length();
            }
            else if (currMovementObj instanceof SliderRepeat_1.SliderRepeat)
                requiredMovement = exports.NORMALISED_RADIUS;
            if (currMovementLength > requiredMovement) {
                currCursorPosition = currCursorPosition.add(currMovement.scale((currMovementLength - requiredMovement) / currMovementLength));
                currMovementLength *= (currMovementLength - requiredMovement) / currMovementLength;
                slider.LazyTravelDistance += currMovementLength;
            }
            if (i === nestedHitObjects.length - 1)
                slider.LazyEndPosition = currCursorPosition;
        }
    }
    getEndCursorPosition(hitObject) {
        let pos = hitObject.StackedPosition;
        if (hitObject instanceof Slider_1.Slider) {
            const slider = hitObject;
            this.computeSliderCursorPosition(slider);
            pos = slider.LazyEndPosition ?? pos;
        }
        return pos;
    }
    ;
    GetDoubleTapness(osuNextObj) {
        if (!osuNextObj)
            return 0;
        const currDeltaTime = Math.max(1, this.DeltaTime);
        const nextDeltaTime = Math.max(1, osuNextObj.DeltaTime);
        const deltaDifference = Math.abs(nextDeltaTime - currDeltaTime);
        const speedRatio = currDeltaTime / Math.max(currDeltaTime, deltaDifference);
        const windowRatio = Math.pow(Math.min(1, currDeltaTime / this.HitWindowGreat), 2);
        return 1 - Math.pow(speedRatio, 1 - windowRatio);
    }
    ;
    OpacityAt(time, hidden) {
        if (time > this.BaseObject.StartTime)
            return 0;
        const fadeInStartTime = this.BaseObject.StartTime - this.BaseObject.TimePreempt;
        const fadeInDuration = this.BaseObject.TimeFadeIn;
        if (hidden) {
            const fadeOutStartTime = this.BaseObject.StartTime - this.BaseObject.TimePreempt + this.BaseObject.TimeFadeIn;
            const fadeOutDuration = this.BaseObject.TimePreempt * 0.3;
            return Math.min((0, helpers_1.clamp)((time - fadeInStartTime) / fadeInDuration, 0, 1), 1 - (0, helpers_1.clamp)((time - fadeOutStartTime) / fadeOutDuration, 0, 1));
        }
        return (0, helpers_1.clamp)((time - fadeInStartTime) / fadeInDuration, 0, 1);
    }
    ;
}
exports.OsuDifficultyHitObject = OsuDifficultyHitObject;
;
