// osu-sr-calculator/dist/Objects/osu/HitObjects/OsuHitObject.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.OsuHitObject = void 0;
const Vector2_1 = require("../../Vector2");
const HitWindows_1 = require("../HitWindows");
const HitObject_1 = require("./HitObject");
const OsuHitWindows_1 = require("./OsuHitWindows");
class OsuHitObject extends HitObject_1.HitObject {
    get Position() { return this.position; }
    ;
    set Position(value) {
        this.position = value;
    }
    ;
    get StackedPosition() { return this.Position.add(this.StackOffset); }
    ;
    get EndPosition() { return this.Position; }
    ;
    get StackedEndPosition() { return this.EndPosition.add(this.StackOffset); }
    ;
    get StackOffset() { return new Vector2_1.Vector2(this.StackHeight * this.Scale * -6.4); }
    ;
    get Radius() { return this.OBJECT_RADIUS * this.Scale; }
    ;
    constructor(osuHitObject) {
        super(osuHitObject);
        this.OBJECT_RADIUS = 64;
        this.OBJECT_DIMENSIONS = new Vector2_1.Vector2(this.OBJECT_RADIUS * 2, this.OBJECT_RADIUS * 2);
        this.BASE_SCORING_DISTANCE = 100;
        this.PREEMPT_MIN = 450;
        this.PREEMPT_MID = 1200;
        this.PREEMPT_MAX = 1800;
        this.TimePreempt = 600;
        this.TimeFadeIn = 400;
        this.StackHeight = 0;
        this.Scale = 0;
        if (osuHitObject.Position)
            this.Position = osuHitObject.Position;
        if (osuHitObject.StackHeight)
            this.StackHeight = osuHitObject.StackHeight;
        if (osuHitObject.Scale)
            this.Scale = osuHitObject.Scale;
    }
    ;
    ApplyDefaultsToSelf(beatmap) {
        super.ApplyDefaultsToSelf(beatmap);
        this.TimePreempt = (0, HitWindows_1.GetDifficultyRange)(beatmap.Difficulty.ApproachRate, this.PREEMPT_MAX, this.PREEMPT_MID, this.PREEMPT_MIN);
        this.TimeFadeIn = 400 * Math.min(1, this.TimePreempt / this.PREEMPT_MIN);
        this.Scale = CalculateScaleFromCircleSize(beatmap.Difficulty.CircleSize, true);
    }
    ;
    CreateHitWindows() {
        return new OsuHitWindows_1.OsuHitWindows();
    }
    ;
}
exports.OsuHitObject = OsuHitObject;
;
const CalculateScaleFromCircleSize = (circleSize, applyFudge = false) => {
    const broken_gamefield_rounding_allowance = 1.00041;
    return (1 - 0.7 * (0, HitWindows_1.GetDifficultyRangeSingle)(circleSize)) / 2 * (applyFudge ? broken_gamefield_rounding_allowance : 1);
};
