// osu-sr-calculator/dist/Objects/osu/HitObjects/Slider.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Slider = void 0;
const SliderTailCircle_1 = require("./SliderObjects/SliderTailCircle");
const SliderTick_1 = require("./SliderObjects/SliderTick");
const OsuHitObject_1 = require("./OsuHitObject");
const SliderHeadCircle_1 = require("./SliderObjects/SliderHeadCircle");
const helpers_1 = require("../../../helpers");
const SliderEventGenerator_1 = require("../../../SliderEventGenerator");
const SliderPath_1 = require("../../../SliderPath");
const PathControlPoint_1 = require("../../../PathControlPoint");
const SliderRepeat_1 = require("./SliderObjects/SliderRepeat");
const Cache_1 = require("../../Cache");
const EmptyHitWindows_1 = require("./EmptyHitWindows");
class Slider extends OsuHitObject_1.OsuHitObject {
    get EndTime() { return this.StartTime + this.SpanCount() * this.Path.Distance / this.Velocity; }
    ;
    get Duration() { return this.EndTime - this.StartTime; }
    ;
    get EndPosition() {
        if (this.endPositionCache.IsValid)
            return this.endPositionCache.Value;
        return this.endPositionCache.Value = this.Position.add(this.CurvePositionAt(1));
    }
    ;
    StackedPositionAt(t) {
        return this.StackedPosition.add(this.CurvePositionAt(t));
    }
    ;
    get Path() { return this.path; }
    ;
    set Path(value) {
        if (!value)
            return;
        this.path.ControlPoints.length = 0;
        this.path.ControlPoints.push(...value.ControlPoints.map(c => new PathControlPoint_1.PathControlPoint(c.Position, c.Type)));
        this.path.ExpectedDistance = value.ExpectedDistance;
    }
    ;
    get Distance() { return this.Path.Distance; }
    ;
    get Position() { return super.Position; }
    set Position(value) {
        super.Position = value;
        this.updateNestedPositions();
    }
    ;
    get RepeatCount() { return this.repeatCount; }
    ;
    set RepeatCount(value) {
        this.repeatCount = value;
        this.updateNestedPositions();
    }
    ;
    get SpanDuration() { return this.Duration / this.SpanCount(); }
    ;
    get ClassicSliderBehaviour() { return this.classicSliderBehaviour; }
    ;
    set ClassicSliderBehaviour(value) {
        this.classicSliderBehaviour = value;
        if (!!this.HeadCircle)
            this.HeadCircle.ClassicSliderBehaviour = value;
        if (!!this.TailCircle)
            this.TailCircle.ClassicSliderBehaviour = value;
    }
    ;
    constructor(slider) {
        super(slider);
        this.endPositionCache = new Cache_1.TypedCache();
        this.path = new SliderPath_1.SliderPath({ ControlPoints: [], OptimiseCatmull: true });
        this.LazyTravelDistance = 0;
        this.LazyTravelTime = 0;
        this.repeatCount = 0;
        this.Velocity = 1;
        this.TickDistance = 0;
        this.TickDistanceMultiplier = 1;
        this.SliderVelocityMultiplier = 1;
        this.Path = slider.Path;
    }
    ;
    SpanCount() {
        return this.RepeatCount + 1;
    }
    ;
    CurvePositionAt(progress) {
        return this.Path.PositionAt(this.ProgressAt(progress));
    }
    ;
    ProgressAt(progress) {
        let p = progress * this.SpanCount() % 1;
        if (this.SpanAt(progress) % 2 === 1)
            p = 1 - p;
        return p;
    }
    ;
    SpanAt(progress) {
        return progress * this.SpanCount();
    }
    ;
    ApplyDefaultsToSelf(beatmap) {
        super.ApplyDefaultsToSelf(beatmap);
        const timingPoint = (0, helpers_1.BinarySearchTimingPoint)(beatmap.TimingPoints, this.StartTime) ?? beatmap.TimingPoints[0];
        this.Velocity = this.BASE_SCORING_DISTANCE * beatmap.Difficulty.SliderMultiplier / (0, helpers_1.GetPrecisionAdjustedBeatLength)(this, timingPoint, "osu");
        const scoringDistance = this.Velocity * timingPoint.BeatLength;
        this.TickDistance = scoringDistance / beatmap.Difficulty.SliderTickRate * this.TickDistanceMultiplier;
    }
    ;
    CreateNestedHitObjects() {
        super.CreateNestedHitObjects();
        const sliderEvents = SliderEventGenerator_1.SliderEventGenerator.Generate(this.StartTime, this.SpanDuration, this.Velocity, this.TickDistance, this.Path.Distance, this.SpanCount());
        for (const e of sliderEvents) {
            switch (e.Type) {
                case SliderEventGenerator_1.SliderEventType.Tick:
                    this.AddNested(new SliderTick_1.SliderTick({
                        SpanIndex: e.SpanIndex,
                        SpanStartTime: e.SpanStartTime,
                        StartTime: e.Time,
                        Position: this.Position.add(this.Path.PositionAt(e.PathProgress)),
                        PathProgress: e.PathProgress,
                        StackHeight: this.StackHeight
                    }));
                    break;
                case SliderEventGenerator_1.SliderEventType.Head:
                    this.HeadCircle = new SliderHeadCircle_1.SliderHeadCircle({
                        StartTime: e.Time,
                        Position: this.Position,
                        StackHeight: this.StackHeight,
                        ClassicSliderBehaviour: this.ClassicSliderBehaviour
                    });
                    this.AddNested(this.HeadCircle);
                    break;
                case SliderEventGenerator_1.SliderEventType.Tail:
                    this.TailCircle = new SliderTailCircle_1.SliderTailCircle({
                        ...this,
                        RepeatIndex: e.SpanIndex,
                        StartTime: e.Time,
                        Position: this.EndPosition,
                        StackHeight: this.StackHeight,
                        ClassicSliderBehaviour: this.ClassicSliderBehaviour,
                        Slider: this
                    });
                    this.AddNested(this.TailCircle);
                    break;
                case SliderEventGenerator_1.SliderEventType.Repeat:
                    this.LastRepeat = new SliderRepeat_1.SliderRepeat({
                        ...this,
                        RepeatIndex: e.SpanIndex,
                        StartTime: this.StartTime + (e.SpanIndex + 1) * this.SpanDuration,
                        Position: this.Position.add(this.Path.PositionAt(e.PathProgress)),
                        StackHeight: this.StackHeight,
                        PathProgress: e.PathProgress,
                        Slider: this
                    });
                    this.AddNested(this.LastRepeat);
                    break;
            }
        }
    }
    ;
    updateNestedPositions() {
        this.endPositionCache.Invalidate();
        for (const nested of this.NestedHitObjects) {
            if (nested instanceof SliderHeadCircle_1.SliderHeadCircle)
                nested.Position = this.Position;
            else if (nested instanceof SliderTailCircle_1.SliderTailCircle)
                nested.Position = this.EndPosition;
            else if (nested instanceof SliderRepeat_1.SliderRepeat)
                nested.Position = this.Position.add(this.Path.PositionAt(nested.PathProgress));
            else if (nested instanceof SliderTick_1.SliderTick)
                nested.Position = this.Position.add(this.Path.PositionAt(nested.PathProgress));
        }
    }
    ;
    CreateHitWindows() {
        return new EmptyHitWindows_1.EmptyHitWindows();
    }
    ;
}
exports.Slider = Slider;
;
